core/app/models/spree/fulfilment_changer.rb

```diff
@@ -0,0 +1,92 @@
+module Spree
+  class FulfilmentChanger
+    include ActiveModel::Validations
+    attr_accessor :current_shipment, :desired_shipment, :variant, :quantity
+
+    def initialize(current_shipment:, desired_shipment:, variant:, quantity:)
+      @current_shipment = current_shipment
+      @desired_shipment = desired_shipment
+      @variant = variant
+      @quantity = quantity
+    end
+
+    validates :quantity, numericality: { greater_than: 0 }
+    validate :current_shipment_not_already_shipped
+    validate :desired_shipment_different_from_current
+    validates :desired_stock_location, presence: true
+    validate :enough_stock_at_desired_location, if: :handle_stock_location_counts?
+
+    def run!
+      return false if invalid?
+      desired_shipment.save! if desired_shipment.new_record?
+
+      current_on_hand_quantity = [current_shipment.inventory_units.on_hand.size, quantity].min
+      new_on_hand_quantity = [desired_shipment.stock_location.count_on_hand(variant), quantity].min
+
+      if handle_stock_location_counts?
```

Comment:
Maybe it makes sense to put all the stock and shipment changing code into a transaction to avoid potential stock count mismatches?